pubsub/gossipsub/gossipsub-v1.1.md

```diff
@@ -634,6 +637,83 @@ In order counter spam that elicits responses and consumes resources, some measur
   further steps and blacklist the peer if the spam persists after the negative score decays.
 
+### Validation Queue Protection
+
+An important aspect of gossipsub is the reliance on validators to signal acceptance of incoming
+messages from the application to the router. The validation is asynchronous, with a typical
+implementation strategy that uses of a front-end queue and a limit to the number of ongoing validations.
+This creates a potential target for attacks, as an attacker can overload the queue by brute force,
+sending spam messages at a very high rate. The effect would be that legitimate messages get dropped
+by the validation front end, resulting in denial of service.
+
+In order to protect the system from this class of attacks, gossipsub v1.1 incorporates a circuit
+breaker that sits before the validation queue and can make informed decisions on whether to
+push a message into the validation queue. This defensive mechanism kicks in when the system detects
+an elevated rate of dropped messages, and makes decisions on whether to accept incoming messages for
+validation based on the statistical performance of peers in the origin IP address. The decision is
+probabilistic and implements a Random Early Drop (RED) strategy that drops messages with a probability
+that depends on the acceptance rates for messages from the origin IP. This strategy can neuter
+attacks on the validation queue, because messages are no longer dropped indiscrimantly in a drop-tail
+fashion.
+
+#### Random Early Drop Algorithm
+
+The algorithm has two aspects:
+- The decision on whether to trigger RED.
+- The decision on whether to drop a message from an origin IP address.
+
+In order to trigger RED, the circuit breaker maintains the following queue statistics:
+- a (decaying) counter for the number of message validations.
+- a (decaying) counter for the number of dropped messages.
+
+The decision on triggering RED is based on comparing the ratio of dropped messages to validations.
+If the ratio exceeds an application configured threshold, then the RED algorithm
+triggers and a decision on whether to accept the message for validation is made based on origin IP
+statistics. There is also a quiet period, such that if no messages have been dropped for a while, the
+circuit breaker turns back off.
+
+In order to make the actual RED decision, the circuit breaker maintains the following statistics per
+IP:
+- a (decaying) counter for the number of accepted messages.
+- a (decaying) counter for the number of duplicate messages, mixed with a weight `W_duplicate`.
+- a (decaying) counter for the number of ignored messages, mixed with a weight `W_ignored`.
+- a (decaying) counter for the number of rejected messages, mixed with a weight `W_rejected`.
+
+The router generates a random float `r` and accepts the message if and only if
+```
+r < (1 + accepted) / (1 + accepted + W_duplicate * duplicate + W_ignored * ignored + W_rejected * rejected)
```

Comment:
It would be great to add recommendations or suggestions of concrete values and reasoning to choose those values (heuristics), so that's it not entirely left to the reader to figure out what a sane set of values would be here.

I wouldn't just plug in the Lotus values here, but maybe the similar weights and explain why. For example:

"With a recipe of W_duplicate=1, W_ignored=2, W_rejected=3, we are rapidly penalising rejections, mildly penalising ignored messages, and (relatively) softly penalising duplicate messages. The result is that clearly misbehaving peers (whose messages lead to outright rejection) will make up for 50% of the decision to break the circuit, while underperforming peers will also factor in, but with less force."

Reply:
I have a defaults section, but yeah we could add some explanation for them.

Reply:
done.